plugins/org.eclipse.xtend.ide/src/org/eclipse/xtend/ide/contentassist/XtendProposalProvider.java

```diff
@@ -133,33 +133,42 @@ public void completeTypeReferenceNoTypeArgs_Type(EObject model, Assignment assig
 		}
 	}
 	
-	@Override
-	protected ITypesProposalProvider.Filter createVisibilityFilter(ContentAssistContext context, int searchFor) {
-		XtendFile file = (XtendFile) context.getRootModel();
-		final char[] contextPackageName = Strings.emptyIfNull(file.getPackage()).toCharArray(); 
-		return new TypeMatchFilters.All(searchFor) {
-			@Override
-			public boolean accept(int modifiers, char[] packageName, char[] simpleTypeName,
-					char[][] enclosingTypeNames, String path) {
-				if (super.accept(modifiers, packageName, simpleTypeName, enclosingTypeNames, path)) {
-					if ("org.eclipse.xtend.lib".equals(String.valueOf(packageName))) {
-						if ("Property".equals(String.valueOf(simpleTypeName))||"Data".equals(String.valueOf(simpleTypeName))) {
-							return false;
-						}
-					}
-					if (Flags.isPublic(modifiers)) {
-						return true;
-					}
-					if (Flags.isPrivate(modifiers)) {
+	private static class VisibilityFilter extends TypeMatchFilters.AbstractFilter {
+		private ContentAssistContext context;
+		public VisibilityFilter(ContentAssistContext context, int searchFor) {
+			super(searchFor);
+			this.context = context;
+		}
+		@Override
+		public boolean accept(int modifiers, char[] packageName, char[] simpleTypeName, char[][] enclosingTypeNames,
+				String path) {
+			if (TypeMatchFilters.isNonInternal(getSearchFor()).accept(modifiers, packageName, simpleTypeName, enclosingTypeNames, path)
+				&& TypeMatchFilters.isAcceptableByPreference().accept(modifiers, packageName, simpleTypeName, enclosingTypeNames, path)) {
+				XtendFile file = (XtendFile) context.getRootModel();
+				final char[] contextPackageName = Strings.emptyIfNull(file.getPackage()).toCharArray();
```

Comment:
Please make 'contextPackageName' a field in the inner class as we'd need to convert the package name to an array for all candidates otherwise.